lib/vmpooler/api/helpers.rb

```diff
@@ -4,6 +4,60 @@ class API
 
     module Helpers
 
+      def protected!
+        return if authorized?
+
+        content_type :json
+
+        result = { 'ok' => false }
+
+        headers['WWW-Authenticate'] = 'Basic realm="Authentication required"'
+        halt 401, JSON.pretty_generate(result)
+      end
+
+      def authorized?
+        @auth ||= Rack::Auth::Basic::Request.new(request.env)
+
+        if @auth.provided? and @auth.basic? and @auth.credentials
+          username, password = @auth.credentials
+
+          if authenticate(Vmpooler::API.settings.config[:auth], username, password)
+            return true
+          end
+        end
+
+        return false
+      end
+
+      def authenticate(auth, username_str, password_str)
+        case auth['provider']
+          when 'ldap'
+            require 'rubygems'
+            require 'net/ldap'
+
+            ldap = Net::LDAP.new(
+              :host => auth[:ldap]['host'],
+              :port => auth[:ldap]['port'] || 389,
```

Comment:
Yeah, it's usually one or the other... straight SSL communication over :636 or start_tls encryption over :389.

In any case this could be more abstracted so that whoever is running the service can choose how they want the vmpooler app to communicate with LDAP.  As-is I'm kind of forcing the second option (start_tls over :389) or double-encryption (start_tls over :636).  :/